### PR TITLE
Implement PoX-4 Locking via Special Contract-Call Handler

### LINKED ISSUE #4090: [Nakamoto] Add `pox_locking::pox_4` lockup methods
The special-case handlers for the PoX-4 lockups need to be implemented.  This is basically a copy/paste of pox_3.rs

### LINKED ISSUE #4091: [Nakamoto] Update `handle_contract_call_special_cases()` in `stackslib/src/clarity_vm/special.rs` to handle PoX-4
The special-case handlers for PoX lockups need to be updated to handle PoX 4, and to handle a read-only PoX-3

### PR DESCRIPTION
**Pull request by jbencin**
### Description

Copy/paste `pox_3.rs` to `pox_4.rs` and rename things as needed. Also, fix `clippy::needless_return` lint

### Applicable issues
- fixes #4090
- fixes #4091

### Checklist
- [ ] Port the following pox3 tests to pox4
  - [ ] `delegate_extend_pox_3`
  - [ ] `delegate_stack_increase`
  - [x] `get_pox_addrs` -> `pox_lock_unlock`
  - [ ] `pox_3_delegate_stx_addr_validation`
  - [ ] `pox_3_getters`
  - [ ] `pox_auto_unlock_ab`
  - [ ] `pox_auto_unlock_ba`
  - [x] `pox_extend_transition`
  - [ ] `simple_pox_lockup_transition_pox_2`
  - [ ] `stack_aggregation_increase`
  - [ ] `stack_increase`
  - [ ] `stack_with_segwit`
- [x] Test that pox3 methods fail (`pox_3_fails`)
- [x] Test that STX locked in pox3 unlocks

### CODE REVIEW

File: pox-locking/src/pox_4.rs
Comment (jcnelson) [inline pox-locking/src/pox_4.rs:153]: duplicate comment?

File: pox-locking/src/pox_4.rs
Comment (jcnelson) [inline pox-locking/src/pox_4.rs:343]: This function `synthesize_pox_2_or_3_event_info()` should be named something more informative, now that pox-4 exists.  Perhaps `synthesize_pox_event_info()`?

Comment (friedger) [reply pox-locking/src/pox_4.rs:343]: I like the name. If we agree on it, I'll update the PR #4094 that changes the pox-lockup::events

Comment (friedger) [reply pox-locking/src/pox_4.rs:343]: Done in https://github.com/stacks-network/stacks-core/pull/4094/commits/915fba782da0009cdef509a0522bdc0476c11265

**Comment by hugocaillard** [comment]
Following this issue on Clarinet: https://github.com/hirosystems/clarinet/issues/1267, we brainstormed with @lgalabru for a clean way to handle pox-4 on Clarinet (regarding the concatenation with `pox-<netowrk>.clar`).

While we all agree that it would be nice to have a standard way to handle such use cases at the Clarity or Clarinet level, it may take a bit of time to get there.

In the meantime, regarding the pox-4 contract specifically, what about leveraging **`is-in-mainnet`** that was introduced in Clarity 2? With the constants defined based on that, it would completely remove the need for contract concatenation. 
```clar
(define-constant PREPARE_CYCLE_LENGTH (if is-in-mainnet u100 u50))
```

**Comment by jbencin** [comment]
@moodmosaic I managed to fix one test `get_pox_addrs()` by doing the following:
- Disable epoch 3.0 and nakamoto blocks. We can test pox4 in Epoch 2.5
- Experiment with stacking start height. We submit the stacking txs at `first_v4_cycle - 3`
- Experiment with when we start checking things, at `first_v4_cycle + 5`

I still don't understand why these block heights work, but the tests clearly show STX are being locked for a couple cycles, then unlocked, so it looks okay to me

Also, going to give you write access to this branch so you can push (since it's in my repo)

File: stackslib/src/chainstate/nakamoto/coordinator/tests.rs
Comment (jcnelson) [inline stackslib/src/chainstate/nakamoto/coordinator/tests.rs:285]: Why did you change this?

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs
Comment (jcnelson) [inline stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:1]: Per our meeting earlier, I think the only things that this file needs to test are:

* PoX 4 instantiates 
* STX locked in PoX 3 unlock automatically
* Once PoX 4 is active, only PoX 4 stacking methods work.  PoX 3 ones fail.

There's no reason to test PoX 2 and PoX behavior, since both these contracts have been defunct for some time.

File: stackslib/src/chainstate/nakamoto/coordinator/tests.rs
Comment (jbencin) [reply stackslib/src/chainstate/nakamoto/coordinator/tests.rs:285]: `test_simple_nakamoto_coordinator_1_tenure_10_blocks()` and `test_simple_nakamoto_coordinator_10_tenures_10_blocks()` started failing as soon as I added support for pox4 in `pox-locking/src/lib.rs`. Aaron suggested this as a way to get the coordinator tests working again

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs
Comment (moodmosaic) [reply stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:1]: @jcnelson
>* PoX 4 instantiates

does [`get_pox_addrs`](https://github.com/stacks-network/stacks-core/pull/4106/files#diff-1b7e3da4025244bdde5f977c36554b94ab89e844c05e4c3c0a106dd9da9c7636R3369) captures this? If it does, I would like to start the current `pox_4_tests.rs` (in this PR) from a clean state and start by just adding this test, and then keep adding to it.

File: stackslib/src/chainstate/nakamoto/coordinator/tests.rs
Comment (jcnelson) [reply stackslib/src/chainstate/nakamoto/coordinator/tests.rs:285]: Ah, okay.  I can see why that is -- now that the special-case handlers are in place, stacking would have locked up all of the STX tied to `peer.config.private_key`, which would prevent this test from running (since it sends STX from this key).  Can you at least use a constant private key so the addresses in this test are deterministic?

Comment (kantai) [inline stackslib/src/chainstate/nakamoto/coordinator/tests.rs:285]: Yep -- I'd recommend using a constant private key, e.g.:

```suggestion
    let private_key = StacksPrivateKey::from_seed(&[2]);
```

File: stackslib/src/chainstate/nakamoto/coordinator/tests.rs
Comment (kantai) [inline stackslib/src/chainstate/nakamoto/coordinator/tests.rs:391]: I would recommend using a constant private key here as well.

**Review by kantai — APPROVED** [review]
LGTM -- I would just recommend using `from_seed` to initialize the necessary new StacksPrivateKeys in `nakamoto::coordinator::tests`

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs
Comment (jcnelson) [inline stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:1]: Missing copyright comment block

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs
Comment (jcnelson) [inline stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:400]: Shouldn't this be `Alice _will not_ auto-unlock`?  Also, wouldn't Alice have unlocked from pox-2 activating?

Comment (kantai) [reply stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:400]: No -- the comment is confusing, but "correct". Alice stacked in pox v1 and so would have experienced auto-unlock but instead she stack-extends in PoX v2.

The v2 unlock height configuration setting _doesn't_ need to match epoch2.1 start (and it doesn't in these tests) -- the auto-unlock occurs after pox-2 is activated, which is why Alice can extend: she's currently locked, and she's extending so she doesn't miss a cycle.

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs
Comment (jcnelson) [inline stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:878]: Can you call this something more descriptive, like "test_pox_3_defunct()`?

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs
Comment (jcnelson) [inline stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:1053]: Can you actually check the reward set itself?

Comment (jbencin) [reply stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:1053]: This line inserts the address into `rewarded` and checks it a few lines later. Is that sufficient or did you mean something else?

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs
Comment (jbencin) [reply stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:878]: Will do

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs
Comment (jcnelson) [reply stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:1053]: That works!  Didn't catch that

**Review by jcnelson — APPROVED** [review]
LGTM; just please address my remaining concerns before merging.  Thanks!

File: stackslib/src/chainstate/nakamoto/coordinator/tests.rs
Comment (jbencin) [reply stackslib/src/chainstate/nakamoto/coordinator/tests.rs]: Used Aaron's suggestion of `StacksPrivateKey::from_seed()`

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs
Comment (jbencin) [reply stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:983]: Trying to figure out how to do this, I think I understand, but want to check:

- Get `StacksTransactionReceipt`s from latest block after executing transactions
- Filter the ones that contain a `StacksTransactionEvent` with a `pox-3` call
- Check that `StacksTransactionReceipt.result` is an error value

Does this sound right?

Comment (jcnelson) [reply stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:983]: Yes; if possible, also test that we specifically fail because contract was defunct (as opposed to some other error that might arise due to a regression in the test).

Comment (kantai) [reply stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:983]: I'm not sure that the `StacksTransactionReceipt`s will be very discerning for these checks -- runtime errors (like these should be) just return a receipt with a result of `err_none`, an empty assetmap, and an empty event list. But, you could usefully assert that a runtime error occurs by checking the receipt.

This is how `pox_3_tests` does this: https://github.com/stacks-network/stacks-core/blob/develop/stackslib/src/chainstate/stacks/boot/pox_3_tests.rs#L512

Comment (kantai) [reply stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:983]: Ah, that's because `pox-3` doesn't seem to actually be marked defunct in these changes. You need to do something like:

```diff
diff --git a/pox-locking/src/lib.rs b/pox-locking/src/lib.rs
index 05303a4d9..057667341 100644
--- a/pox-locking/src/lib.rs
+++ b/pox-locking/src/lib.rs
@@ -107,6 +107,19 @@ pub fn handle_contract_call_special_cases(
             result,
         );
     } else if *contract_id == boot_code_id(POX_3_NAME, global_context.mainnet) {
+        if !pox_3::is_read_only(function_name) && global_context.epoch_id >= StacksEpochId::Epoch25
+        {
+            warn!("PoX-3 function call attempted on an account after Epoch 2.5";
+                  "v3_unlock_ht" => global_context.database.get_v3_unlock_height(),
+                  "current_burn_ht" => global_context.database.get_current_burnchain_block_height(),
+                  "function_name" => function_name,
+                  "contract_id" => %contract_id
+            );
+            return Err(ClarityError::Runtime(
+                RuntimeErrorType::DefunctPoxContract,
+                None,
+            ));
+        }
         return pox_3::handle_contract_call(
             global_context,
             sender,
diff --git a/pox-locking/src/pox_3.rs b/pox-locking/src/pox_3.rs
index cccfbb264..e0110987e 100644
--- a/pox-locking/src/pox_3.rs
+++ b/pox-locking/src/pox_3.rs
@@ -32,6 +32,33 @@ use crate::pox_2::{parse_pox_extend_result, parse_pox_increase, parse_pox_stacki
 use crate::{LockingError, POX_3_NAME};
 
 /////////////////////// PoX-3 /////////////////////////////////
+/// is a PoX-3 function call read only?
+pub(crate) fn is_read_only(func_name: &str) -> bool {
+    "get-pox-rejection" == func_name
+        || "is-pox-active" == func_name
+        || "burn-height-to-reward-cycle" == func_name
+        || "reward-cycle-to-burn-height" == func_name
+        || "current-pox-reward-cycle" == func_name
+        || "get-stacker-info" == func_name
+        || "get-check-delegation" == func_name
+        || "get-reward-set-size" == func_name
+        || "next-cycle-rejection-votes" == func_name
+        || "get-total-ustx-stacked" == func_name
+        || "get-reward-set-pox-address" == func_name
+        || "get-stacking-minimum" == func_name
+        || "check-pox-addr-version" == func_name
+        || "check-pox-addr-hashbytes" == func_name
+        || "check-pox-lock-period" == func_name
+        || "can-stack-stx" == func_name
+        || "minimal-can-stack-stx" == func_name
+        || "get-pox-info" == func_name
+        || "get-delegation-info" == func_name
+        || "get-allowance-contract-callers" == func_name
+        || "get-num-reward-set-pox-addresses" == func_name
+        || "get-partial-stacked-by-cycle" == func_name
+        || "get-total-pox-rejection" == func_name
+}
+
 
 /// Lock up STX for PoX for a time.  Does NOT touch the account nonce.
 pub fn pox_lock_v3(
```

Comment (jbencin) [reply stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:983]: Thanks @kantai, that fixed it! Will check in and merge this ASAP

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs
Comment (kantai) [inline stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:978]: Can you make sure that every transaction failed, i.e. `assert_eq!(receipts.len(), txs.len()`?

File: stackslib/src/chainstate/stacks/boot/pox_4_tests.rs
Comment (kantai) [inline stackslib/src/chainstate/stacks/boot/pox_4_tests.rs:975]: Can you check that the `result` is specifically `(err none)`?

**Review by kantai — APPROVED** [review]
LGTM, just a couple tweaks to the test case.

**Comment by rondales** [comment]
> ### Description
> Copy/paste `pox_3.rs` to `pox_4.rs` and rename things as needed. Also, fix `clippy::needless_return` lint
> 
> ### Applicable issues
> * fixes [[Nakamoto] Add `pox_locking::pox_4` lockup methods #4090](https://github.com/stacks-network/stacks-core/issues/4090)
> * fixes [[Nakamoto] Update `handle_contract_call_special_cases()` in `stackslib/src/clarity_vm/special.rs` to handle PoX-4 #4091](https://github.com/stacks-network/stacks-core/issues/4091)
> 
> ### Additional info (benefits, drawbacks, caveats)
> ### Checklist
> * [ ]  Port the following pox3 tests to pox4
>   
>   * [ ]  `delegate_extend_pox_3`
>   * [ ]  `delegate_stack_increase`
>   * [x]  `get_pox_addrs` -> `pox_lock_unlock`
>   * [ ]  `pox_3_delegate_stx_addr_validation`
>   * [ ]  `pox_3_getters`
>   * [ ]  `pox_auto_unlock_ab`
>   * [ ]  `pox_auto_unlock_ba`
>   * [x]  `pox_extend_transition`
>   * [ ]  `simple_pox_lockup_transition_pox_2`
>   * [ ]  `stack_aggregation_increase`
>   * [ ]  `stack_increase`
>   * [ ]  `stack_with_segwit`
> * [x]  Test that pox3 methods fail (`pox_3_fails`)
> * [x]  Test that STX locked in pox3 unlocks

**Pull request merged**
